File: bench/gain/holo.cpp
// File: holo.cpp
// Project: gain
// Created Date: 12/12/2022
// Author: Shun Suzuki
// -----
// Last Modified: 31/01/2023
// Modified By: Shun Suzuki (suzuki@hapis.k.u-tokyo.ac.jp)
// -----
// Copyright (c) 2022 Shun Suzuki. All rights reserved.
//

#if _MSC_VER
#pragma warning(push)
#pragma warning(disable : 26495)
#endif
#if defined(__GNUC__) && !defined(__llvm__)
#pragma GCC diagnostic push
#endif
#ifdef __clang__
#pragma clang diagnostic push
#endif
#include <benchmark/benchmark.h>
#if _MSC_VER
#pragma warning(pop)
#endif
#if defined(__GNUC__) && !defined(__llvm__)
#pragma GCC diagnostic pop
#endif
#ifdef __clang__
#pragma clang diagnostic pop
#endif

#include <autd3/gain/holo.hpp>
#include <random>

#include "../utils.hpp"

constexpr size_t FOCI_SIZE = 10;

static std::vector<autd3::Vector3> gen_foci(const autd3::Vector3& center, const size_t n, const autd3::driver::autd3_float_t rx,
                                            const autd3::driver::autd3_float_t ry, const autd3::driver::autd3_float_t rz, const int32_t seed = 0) {
  std::vector<autd3::Vector3> foci;
  foci.reserve(n);

  std::mt19937 mt(seed);
  std::normal_distribution norm_x(-rx, rx);
  std::normal_distribution norm_y(-ry, ry);
  std::normal_distribution norm_z(-rz, rz);

  for (size_t i = 0; i < n; i++) foci.emplace_back(center + autd3::Vector3(norm_x(mt), norm_y(mt), norm_z(mt)));

  return foci;
}

static void bm_gain_holo_sdp(benchmark::State& state) {
  const auto geometry = setup_geometry(state.range(0), state.range(1));

  const autd3::Vector3 center = geometry.center() + autd3::Vector3(0, 0, 300);

  const auto foci = gen_foci(center, FOCI_SIZE, 100, 100, 100);

  const auto backend = autd3::gain::holo::EigenBackend::create();
  autd3::gain::holo::SDP g(backend);

  for (auto& focus : foci) g.add_focus(focus, 1);

  for (auto _ : state) {
    g.calc(geometry);
  }
}

static void bm_gain_holo_evd(benchmark::State& state) {
  const auto geometry = setup_geometry(state.range(0), state.range(1));

  const autd3::Vector3 center = geometry.center() + autd3::Vector3(0, 0, 300);

  const auto foci = gen_foci(center, FOCI_SIZE, 100, 100, 100);

  const auto backend = autd3::gain::holo::EigenBackend::create();
  autd3::gain::holo::EVD g(backend);

  for (auto& focus : foci) g.add_focus(focus, 1);

  for (auto _ : state) {
    g.calc(geometry);
  }
}

static void bm_gain_holo_lss(benchmark::State& state) {
  const auto geometry = setup_geometry(state.range(0), state.range(1));

  const autd3::Vector3 center = geometry.center() + autd3::Vector3(0, 0, 300);

  const auto foci = gen_foci(center, FOCI_SIZE, 100, 100, 100);

  const auto backend = autd3::gain::holo::EigenBackend::create();
  autd3::gain::holo::LSS g(backend);

  for (auto& focus : foci) g.add_focus(focus, 1);

  for (auto _ : state) {
    g.calc(geometry);
  }
}

static void bm_gain_holo_gs(benchmark::State& state) {
  const auto geometry = setup_geometry(state.range(0), state.range(1));

  const autd3::Vector3 center = geometry.center() + autd3::Vector3(0, 0, 300);

  const auto foci = gen_foci(center, FOCI_SIZE, 100, 100, 100);

  const auto backend = autd3::gain::holo::EigenBackend::create();
  autd3::gain::holo::GS g(backend);

  for (auto& focus : foci) g.add_focus(focus, 1);

  for (auto _ : state) {
    g.calc(geometry);
  }
}

static void bm_gain_holo_gspat(benchmark::State& state) {
  const auto geometry = setup_geometry(state.range(0), state.range(1));

  const autd3::Vector3 center = geometry.center() + autd3::Vector3(0, 0, 300);

  const auto foci = gen_foci(center, FOCI_SIZE, 100, 100, 100);

  const auto backend = autd3::gain::holo::EigenBackend::create();
  autd3::gain::holo::GSPAT g(backend);

  for (auto& focus : foci) g.add_focus(focus, 1);

  for (auto _ : state) {
    g.calc(geometry);
  }
}

static void bm_gain_holo_lm(benchmark::State& state) {
  const auto geometry = setup_geometry(state.range(0), state.range(1));

  const autd3::Vector3 center = geometry.center() + autd3::Vector3(0, 0, 300);

  const auto foci = gen_foci(center, FOCI_SIZE, 100, 100, 100);

  const auto backend = autd3::gain::holo::EigenBackend::create();
  autd3::gain::holo::LM g(backend);

  for (auto& focus : foci) g.add_focus(focus, 1);

  for (auto _ : state) {
    g.calc(geometry);
  }
}

static void bm_gain_holo_greedy(benchmark::State& state) {
  const auto geometry = setup_geometry(state.range(0), state.range(1));

  const autd3::Vector3 center = geometry.center() + autd3::Vector3(0, 0, 300);

  const auto foci = gen_foci(center, FOCI_SIZE, 100, 100, 100);

  const auto backend = autd3::gain::holo::EigenBackend::create();
  autd3::gain::holo::Greedy g(backend);

  for (auto& focus : foci) g.add_focus(focus, 1);

  for (auto _ : state) {
    g.calc(geometry);
  }
}

BENCHMARK(bm_gain_holo_sdp)->Args({1, 1})->Args({2, 2})->Args({3, 3})->Args({4, 4});
BENCHMARK(bm_gain_holo_evd)->Args({1, 1})->Args({2, 2})->Args({3, 3})->Args({4, 4});
BENCHMARK(bm_gain_holo_lss)->Args({1, 1})->Args({2, 2})->Args({3, 3})->Args({4, 4});
BENCHMARK(bm_gain_holo_gs)->Args({1, 1})->Args({2, 2})->Args({3, 3})->Args({4, 4});
BENCHMARK(bm_gain_holo_gspat)->Args({1, 1})->Args({2, 2})->Args({3, 3})->Args({4, 4});
BENCHMARK(bm_gain_holo_lm)->Args({1, 1})->Args({2, 2})->Args({3, 3})->Args({4, 4});
BENCHMARK(bm_gain_holo_greedy)->Args({1, 1})->Args({2, 2})->Args({3, 3})->Args({4, 4});

File: rust/autd3-examples/src/tests/holo.rs
/*
 * File: holo.rs
 * Project: tests
 * Created Date: 29/05/2021
 * Author: Shun Suzuki
 * -----
 * Last Modified: 11/05/2023
 * Modified By: Shun Suzuki (suzuki@hapis.k.u-tokyo.ac.jp)
 * -----
 * Copyright (c) 2021 Shun Suzuki. All rights reserved.
 *
 */

#[macro_export]
macro_rules! holo {
    ($autd:ident) => {{
        use autd3_gain_holo::*;

        $autd.send(SilencerConfig::default())?;

        let m = Sine::new(150);

        let center = $autd.geometry().center() + Vector3::new(0., 0., 150.0);
        let p = Vector3::new(30., 0., 0.);

        println!("[0]: SDP");
        println!("[1]: EVP");
        println!("[2]: GS");
        println!("[3]: GSPAT");
        println!("[4]: LSS");
        println!("[5]: LM");
        println!("[6]: Greedy");
        println!("[Others]: GS-PAT");
        print!("{}", "Choose number: ".green().bold());
        io::stdout().flush()?;

        let mut s = String::new();
        io::stdin().read_line(&mut s)?;
        match s.trim().parse::<usize>() {
            Ok(0) => {
                let mut g = SDP::<NalgebraBackend>::new();
                g.add_focus(center + p, 1.);
                g.add_focus(center - p, 1.);
                $autd.send((m, g))?;
            }
            Ok(1) => {
                let mut g = EVP::<NalgebraBackend>::new();
                g.add_focus(center + p, 1.);
                g.add_focus(center - p, 1.);
                $autd.send((m, g))?;
            }
            Ok(2) => {
                let mut g = GS::<NalgebraBackend>::new();
                g.add_focus(center + p, 1.);
                g.add_focus(center - p, 1.);
                $autd.send((m, g))?;
            }
            Ok(3) => {
                let mut g = GSPAT::<NalgebraBackend>::new();
                g.add_focus(center + p, 1.);
                g.add_focus(center - p, 1.);
                $autd.send((m, g))?;
            }
            Ok(4) => {
                let mut g = LSS::<NalgebraBackend>::new();
                g.add_focus(center + p, 1.);
                g.add_focus(center - p, 1.);
                $autd.send((m, g))?;
            }
            Ok(5) => {
                let mut g = LM::<NalgebraBackend>::new();
                g.add_focus(center + p, 1.);
                g.add_focus(center - p, 1.);
                $autd.send((m, g))?;
            }
            Ok(6) => {
                let mut g = Greedy::new();
                g.add_focus(center + p, 1.);
                g.add_focus(center - p, 1.);
                $autd.send((m, g))?;
            }
            _ => {
                let mut g = GSPAT::<NalgebraBackend>::new();
                g.add_focus(center + p, 1.);
                g.add_focus(center - p, 1.);
                $autd.send((m, g))?;
            }
        };
    }};
}
